pkg/webhook/conversion/chain.go

```diff
@@ -0,0 +1,251 @@
+package conversion
+
+import (
+	"strings"
+
+	"github.com/flant/shell-operator/pkg/utils/string_helper"
+)
+
+type ChainStorage map[string]*Chain
+
+func NewChainStorage() ChainStorage {
+	return make(map[string]*Chain)
+}
+
+type Chain struct {
+	// Index: ruleID ("srcVer->desiredVer") to a sequence of "primitive" ruleIDs.
+	PathsCache map[string][]string
+	// from -> to cache
+	FromToCache map[string]map[string]bool
+}
+
+// Access
+
+func (cs ChainStorage) Get(crdName string) *Chain {
+	if _, ok := cs[crdName]; !ok {
+		cs[crdName] = &Chain{
+			PathsCache:  make(map[string][]string),
+			FromToCache: make(map[string]map[string]bool),
+		}
+	}
+	return cs[crdName]
+}
+
+func (c *Chain) Put(rule ConversionRule) {
+	// paths "from->to"
+	ruleID := rule.String()
+	c.PathsCache[ruleID] = []string{ruleID}
+	// from -> to
+	if _, ok := c.FromToCache[rule.FromVersion]; !ok {
+		c.FromToCache[rule.FromVersion] = make(map[string]bool)
+	}
+	c.FromToCache[rule.FromVersion][rule.ToVersion] = true
+}
+
+// Calculations
+
+func (cs ChainStorage) FindConversionChain(crdName string, rule ConversionRule) []string {
+	chain, ok := cs[crdName]
+	if !ok {
+		return nil
+	}
+
+	// Return if there is no path to a desired version, trimmed or full.
+	if !chain.HasTargetVersion(rule.ToVersion) {
+		return nil
+	}
+
+	for {
+		p := chain.SearchPathForRule(rule)
+		if len(p) > 0 {
+			return p
+		}
+
+		// Fill cache with more paths.
+		newPaths := map[string][]string{}
+
+		// Try only ids that starts from a source version.
+		for _, ruleIDToCheck := range chain.IDsByFromVersion(rule) {
+			ruleToCheck := RuleFromString(ruleIDToCheck)
+
+			if ruleToCheck.ShortToVersion() == rule.ShortFromVersion() {
+				// Ignore loops.
+				continue
+			}
+
+			// toVersion in ruleIDToCheck is a new start. Get toVersions available starting from it.
+			for _, nextRule := range chain.NextRules(ruleToCheck.ToVersion) {
+				newRule := ConversionRule{
+					FromVersion: rule.FromVersion,
+					ToVersion:   nextRule.ToVersion,
+				}
+
+				if newRule.ShortToVersion() == rule.ShortFromVersion() {
+					// Ignore loops.
+					continue
+				}
+
+				newPath := append(chain.PathsCache[ruleIDToCheck], nextRule.String())
+
+				// This path is already discovered.
+				p := chain.SearchPathForRule(newRule)
+				if len(p) != 0 {
+					continue
+				}
+
+				newPaths[newRule.String()] = newPath
+			}
+		}
+
+		// break if no new paths are discovered.
+		if len(newPaths) == 0 {
+			break
+		}
+
+		// Put new paths in cache.
+		for id, path := range newPaths {
+			chain.PathsCache[id] = path
+		}
+	}
+
+	return nil
+}
+
+func (c Chain) SearchPathForRule(rule ConversionRule) []string {
+	IDs := []string{}
```

Comment:
`ids`

Reply:
Done. No more string ids, Rule struct is used as key.